gtsfm/runner/gtsfm_runner_base.py

```diff
@@ -86,20 +96,39 @@ def construct_scene_optimizer(self) -> SceneOptimizer:
 
         return scene_optimizer
 
+    def construct_retriever(self):
+        """Set up retriever module."""
+        matching_regime = ImageMatchingRegime(self.parsed_args.matching_regime)
+
+        if matching_regime == ImageMatchingRegime.EXHAUSTIVE:
+            retriever = ExhaustiveRetriever()
+
+        elif matching_regime == ImageMatchingRegime.SEQUENTIAL:
+            retriever = SequentialRetriever(max_frame_lookahead=self.parsed_args.max_frame_lookahead)
+
+        return retriever
+
     def run(self) -> None:
+        """Run the SceneOptimizer."""
         start_time = time.time()
+
+        # create dask client
+        cluster = LocalCluster(
+            n_workers=self.parsed_args.num_workers, threads_per_worker=self.parsed_args.threads_per_worker
+        )
+
+        pairs_graph = self.retriever.create_computation_graph(self.loader)
+        with Client(cluster), performance_report(filename="dask-report.html"):
```

Comment:
@johnwlambert why do we now have a second `with Client` clause here? The compute only happens in line 135, right?

Reply:
Hi @dellaert, the graph is now partitioned into 2 separate computation graphs.

There is a `.compute()` call on line 122, and also on 135. Dask cannot iterate over a sequence of unknown length, which the image pair indices end up being after the Retriever.

I don't think there is any way around this, unfortunately, in our Dask paradigm.